tests/s25Main/UI/testImageButton.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (C) 2005 - 2022 Settlers Freaks (sf-team at siedler25.org)
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "controls/ctrlImageButton.h"
+#include "ogl/ITexture.h"
+#include "uiHelper/uiHelpers.hpp"
+#include <rttr/test/random.hpp>
+#include <turtle/mock.hpp>
+#include <boost/test/unit_test.hpp>
+
+using namespace rttr::test;
+
+namespace {
+MOCK_BASE_CLASS(TestTexture, ITexture)
+{
+public:
+    TestTexture(Extent size) : size_(size) {}
+    Position GetOrigin() const override
+    {
+        return Position::all(0);
```

Comment:
will do. I've started doing it and found some problems. I don't quite get the idea of image origin - I assume this should generally work so that if I set origin to (5,8) and draw image somewhere in the game at (30,40), then at this position the pixel from the (5,8) is drawn and the image's top left corner starts at (25,32).

This is problematic when drawing image buttons - my assumption is that they should always be centered in the buttons, regardless of the origin values. Should this be the case or am I getting something wrong?

Reply:
Yes IIRC the origin is mostly used for the textures of buildings etc where the origin may vary depending on the nation etc. I don't think it is useful for button textures, which I mentioned in https://github.com/Return-To-The-Roots/s25client/pull/1534/files#r973562992
So yes I do share your opinion that the image should always be centred and the origin ignored for buttons.

Maybe @Spikeone has something to add.

Reply:
I've had a slight brain meltdown because unit tests returned for me different position values passed to `img.Draw()` when specifying different origins, but then I remembered that the origin thingy is resolved inside that function ( https://github.com/Return-To-The-Roots/s25client/blob/master/libs/s25main/ogl/glArchivItem_Bitmap.cpp#L45 ) so I think it has to be that way. :)

Reply:
Ok, that means that the button control MUST not use the origin when calling `Draw` as that is handled by `Draw`.
Or actually if `Draw` handles `origin` then in order to have the image always centered the button control must **add** the origin to account for it being removed in the draw call. So the current change at https://github.com/Return-To-The-Roots/s25client/pull/1536/files#diff-adfa4d0f1a6924716e377c3f74e76a930efc15607563954cee41631febf87c4fR14 would make the image being offset by 2 times the origin.

I guess current button textures don't have an origin set so this wasn't an issue before.

Reply:
I you look closely then GetImageRect() creates rect with `position = -origin` so the calculations are correct there, because my subtraction effectively adds ;) I guess that's confusing though, so I'll look into getting origin directly and add it there instead.

Reply:
That would make a great end-to-end test: Specify a non-zero origin for an image used in a button and then check that the actual draw call (i.e. the OpenGL coords) will draw the image at the exact coordinates. I.e. that you modification and the origin handling in the draw call actually do cancel each other out as you describe here (I checked: you are right)
And maybe add a short comment to the line removing the origin that this "compensates" (or any better word) the offset so the image is drawn exactly where specified.

Reply:
I've updated the comment, doing the end-to-end test is yet to follow.

Reply:
> Yes IIRC the origin is mostly used for the textures of buildings etc where the origin may vary depending on the nation etc. I don't think it is useful for button textures, which I mentioned in https://github.com/Return-To-The-Roots/s25client/pull/1534/files#r973562992 So yes I do share your opinion that the image should always be centred and the origin ignored for buttons.
> 
> Maybe @Spikeone has something to add.

Only if we plan to add buttons with text and image - but currently I don't think we'd need that in any of the current interfaces.